src/docbuild/utils/pidlock.py

```diff
@@ -5,165 +5,169 @@
-                    
-                    # Recursively call acquire() to try and grab the lock immediately.
-                    return self.acquire() 
-
-            except FileNotFoundError:
-                # Race condition: another process removed the stale lock before us. Retry.
-                return self.acquire() 
-            except ValueError:
-                # Invalid PID format. Treat as stale and clean up.
-                log.warning(f"Lock file at {self.lock_path} contains invalid PID. Removing and retrying.")
+
+                # Stale lock: clean and retry
+                log.warning(
+                    f"Found stale lock file at {self.lock_path} (PID {pid}). Removing and retrying acquisition."
+                )
                 self.lock_path.unlink()
-                return self.acquire()
+
+            except (FileNotFoundError, ValueError):
+                log.warning(
+                    f"Lock file at {self.lock_path} missing or invalid. Removing and retrying."
+                )
+                self.lock_path.unlink(missing_ok=True)
+
             except OSError as e:
-                # Non-critical I/O error during read/unlink attempt
                 log.error(f"Non-critical error while checking lock file: {e}")
-            
-        # If all checks and retries fail to acquire the lock, raise a final error.
-        raise RuntimeError(f"Failed to acquire lock for {self.resource_path} after multiple checks.")
+                try:
+                    self.lock_path.unlink(missing_ok=True)
+                except Exception:
+                    pass
+                # Attempt one last atomic creation
+                try:
+                    lock_fd = os.open(
+                        self.lock_path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o644
+                    )
+                    with os.fdopen(lock_fd, "w") as f:
+                        f.write(f"{current_pid}\n")
+                    self._lock_acquired = True
+                    self.lock_file = self.lock_path
+                    atexit.register(self.release)
+                    return
+                except Exception as e2:
+                    raise RuntimeError(
+                        f"Failed to recover from non-critical lock read error: {e2}"
+                    )
 
     def release(self) -> None:
         """Release the lock file."""
         if self._lock_acquired and self.lock_path.exists():
             try:
                 self.lock_path.unlink()
                 self._lock_acquired = False
+                self.lock_file = None
                 log.debug("Released lock at %s.", self.lock_path)
-                
-                # Unregister the cleanup function
-                atexit.unregister(self.release)
+                try:
+                    atexit.unregister(self.release)
+                except (ValueError, AttributeError):
+                    log.debug("atexit.unregister failed or unavailable.")
```

Comment:
Maybe the order is incorrect? Shouldn't we _first_ unregister the cleanup function, and _then_ remove the file? 🤔 


```suggestion
                try:
                    atexit.unregister(self.release)
                # I'm not sure about catching these:
                except (ValueError, AttributeError):
                    log.debug("atexit.unregister failed or unavailable.")
                self.lock_path.unlink()
                self._lock_acquired = False
                self.lock_file = None
                log.debug("Released lock at %s.", self.lock_path)
```